src/monitors/BytecodeProfilingMonitor.v3

```diff
@@ -0,0 +1,423 @@
+		}
+	}
+
+	private def beforeMemRead(profile_state: SourceProfileState, opcode_read_pc: int,
+							  dynamicLoc: DynamicLoc, mem: Memory, address: u64, size: u64) -> Resumption {
+		if (dynamicLoc.pc != opcode_read_pc) {
+			// check for opcode operand read.
+			for (i: u32 = 0; i < size; i++) {
+				var b = mem.read_u8(u32.!(address + i), 0).result;
+				profile_state.try_record_operand(u32.!(address + i), b);
+			}
+		} else { // opcode read
+			var b = mem.read_u8(u32.!(address), 0).result;
+			profile_state.step(u32.!(address), b);
+		}
+		return Resumption.Continue;
+	}
+
+	private def beforeFuncExecOrReturn(dynamicLoc: DynamicLoc, is_enter: bool) -> Resumption {
+		var func_id = dynamicLoc.func.decl.func_index;
+		for (dispatches = potential_dispatch_functions; dispatches != null; dispatches = dispatches.tail) {
+			var state = dispatches.head;
+			if (state.dispatch_func_id == func_id) {
+				if (is_enter) {
+					state.sourceFuncStart();
+				} else {
+					state.sourceFuncReturn();
+				}
+			} else {
+				state.wasmFuncEvent(func_id, is_enter);
+			}
+		}
+		return Resumption.Continue;
+	}
+}
+
+// iterates over all wasm functions to detect possible dispatch functions using simple heuristics.
+private class DispatchFunctionBytecodeVisitor(bi: BytecodeIterator) extends BytecodeVisitor {
+	var opcode_read_pc = -1;
+	var is_dispatch_func = false;
+
+	def visit_BR_TABLE(labels: Range<u31>) {
+		if (!is_dispatch_func && labels.length >= MIN_DISPATCH_CASE_COUNT && opcode_read_pc >= 0) {
+			is_dispatch_func = true;
+		}
+	}
+	def visit_I32_LOAD8_U(imm: MemArg) {
+		if (!is_dispatch_func) {
+			opcode_read_pc = bi.pc;
+		}
+	}
+
+	def reset() {
+		opcode_read_pc = -1;
+		is_dispatch_func = false;
+	}
+}
+
+private class SourceProfileState(count_probe: FuelCountProbe, dispatch_func_id: int, output_prefix: string) {
+	var call_stack = ListStack<u32>.new();
+	// complete execution trace.
+	var trace: List<TraceEntry>;
+	// the recovered bytecode for each source function.
+	var recovered_bytecodes = HashMap<u32, HashMap<u32, u8>>.new(int.!<u32>, u32.==);
+	// address of the last read bytecode.
+	var last_opcode_addr: u32;
+	// total costs of each bytecode execution.
+	def hotness: BytecodeHotness = BytecodeHotness.new(output_prefix);
+	var last_probe_event: ProbeEvent = ProbeEvent.RETURN;
+	var last_step_count: u32;
+	var last_probe_event_count: u32;
+	var cur_wasm_func_trace: List<WasmFuncTrace>;
+
+	// updates profiler state upon wasm function invocation/return, during the execution of a
+	// valid source opcode. This is used to understand which wasm functions are associated to
+	// particular opcodes.
+	def wasmFuncEvent(wasm_func_id: int, is_start: bool) {
+		var wasm_trace = WasmFuncTrace(wasm_func_id, u32.!(count_probe.count) - last_step_count, is_start);
+		cur_wasm_func_trace = List.new(wasm_trace, cur_wasm_func_trace);
+	}
+
+	// updates the profiler state upon source function invocation.
+	def sourceFuncStart() {
```

Comment:
We should pick a good name to refer to the higher-level functions and bytecode. What about "guest function" and "guest bytecode"?

Reply:
done